core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageInputStreamTest.java

```diff
@@ -0,0 +1,307 @@
+        final InputStream inputStream = streamMessage
+                .asInputStream(x -> HttpData.wrap(x.toString().getBytes()));
+        final byte[] expected = ImmutableList.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void readWithStreamMessageOperators() throws Exception {
+        final StreamMessage<Integer> streamMessage = StreamMessage
+                .of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+                .filter(x -> x % 2 == 0)
+                .map(x -> x + 10); // 12, 14, 16, 18, 20
+        final InputStream inputStream = streamMessage
+                .asInputStream(x -> HttpData.wrap(x.toString().getBytes()));
+        final byte[] expected = ImmutableList.of(12, 14, 16, 18, 20)
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void readWithOffset() throws Exception {
+        final StreamMessage<byte[]> streamMessage = StreamMessage.of(new byte[] {1, 2, 3, 4, 5});
+        final InputStream inputStream = streamMessage.asInputStream(HttpData::wrap);
+
+        final byte[] result = new byte[5];
+        final int len = inputStream.read(result, 1, 3);
+
+        assertThat(len).isEqualTo(3);
+        assertThat(result).isEqualTo(new byte[] { 0, 1, 2, 3, 0});
+        assertThat(inputStream.available()).isEqualTo(2);
+    }
+
+    @Test
+    void readByMultipleThreads_shouldWaitSubscribed() throws Exception {
+        final Publisher<Integer> publisher = Flux.range(1, 5)
+                                                 .delaySubscription(Duration.ofSeconds(3));
```

Comment:
(FYI) To test that all threads are waiting `upstream.whenSubscribed` on `read`,
I add `delaySubscription(3 sec)`.